src/app-layer-htp-range.c

```diff
@@ -0,0 +1,422 @@
+{
+    THashShutdown(ContainerUrlRangeList.ht);
+}
+
+uint32_t HttpRangeContainersTimeoutHash(struct timeval *ts)
+{
+    uint32_t cnt = 0;
+
+    for (size_t i = 0; i < ContainerUrlRangeList.ht->config.hash_size; i++) {
+        THashHashRow *hb = &ContainerUrlRangeList.ht->array[i];
+
+        if (HRLOCK_TRYLOCK(hb) != 0)
+            continue;
+        /* hash bucket is now locked */
+        THashData *h = hb->head;
+        while (h) {
+            THashData *n = h->next;
+            if (ContainerValueRangeTimeout(h->data, ts)) {
+                /* remove from the hash */
+                if (h->prev != NULL)
+                    h->prev->next = h->next;
+                if (h->next != NULL)
+                    h->next->prev = h->prev;
+                if (hb->head == h)
+                    hb->head = h->next;
+                if (hb->tail == h)
+                    hb->tail = h->prev;
+                h->next = NULL;
+                h->prev = NULL;
+                HttpRangeContainerFile *c = h->data;
+                SCMutexLock(&c->mutex);
+                FileCloseFile(c->files, NULL, 0, FILE_TRUNCATED);
+                // we should log the timed out file somehow...
+                RangeContainerFree(c);
+                SCMutexUnlock(&c->mutex);
+                THashDataMoveToSpare(ContainerUrlRangeList.ht, h->data);
+            }
+            h = n;
+        }
+        HRLOCK_UNLOCK(hb);
+    }
+
+    return cnt;
+}
+
+void *HttpRangeContainerUrlGet(const uint8_t *key, size_t keylen, struct timeval *ts)
+{
+    HttpRangeContainerFile lookup;
+    // cast so as not to have const in the structure
+    lookup.key = (uint8_t *)key;
+    lookup.len = keylen;
+    struct THashDataGetResult res = THashGetFromHash(ContainerUrlRangeList.ht, &lookup);
+    if (res.data) {
+        // nothing more to do if (res.is_new)
+        ContainerUrlRangeUpdate(res.data->data, ts->tv_sec + ContainerUrlRangeList.timeout);
+        THashDecrUsecnt(res.data);
+        HttpRangeContainerFile *c = res.data->data;
+        if (c->nbref == UINT16_MAX) {
+            THashDataUnlock(res.data);
+            return NULL;
+        }
+        c->nbref++;
+        THashDataUnlock(res.data);
+        return res.data->data;
+    }
+    return NULL;
+}
+
+HttpRangeContainerBlock *ContainerUrlRangeOpenFile(HttpRangeContainerFile *c, uint64_t start,
+        uint64_t end, uint64_t total, const StreamingBufferConfig *sbcfg, const uint8_t *name,
+        uint16_t name_len, uint16_t flags)
+{
+    SCMutexLock(&c->mutex);
+    if (c->files->tail == NULL) {
+        if (FileOpenFileWithId(c->files, sbcfg, 0, name, name_len, NULL, 0, flags) != 0) {
+            SCLogDebug("open file for range failed");
+            c->nbref--;
+            SCMutexUnlock(&c->mutex);
+            return NULL;
+        }
+    }
+    HttpRangeContainerBlock *curf = SCCalloc(1, sizeof(HttpRangeContainerBlock));
+    if (curf == NULL) {
+        c->nbref--;
```

Comment:
Good question.

Maybe `FileOpenFileWithId ` should rather be done in `ContainerUrlRangeSet` right after `FileContainerAlloc()` as it is meant to be done only once per `HttpRangeContainerFile` on the opening of the first range.
But it will be a bit painful to transport all the arguments needed by `FileOpenFileWithId` up to `ContainerUrlRangeSet`
What do you think ?

We can also exchange and try to allocate `curf` before checking `if (c->files->tail == NULL)`